### PR TITLE
Add cursorWidth and cursorRadius to TextField (Material) cursor

### PR DESCRIPTION
**Pull request by sandrasandeep**
The cursor in TextField is currently unlike native iOS and Material cursors. This PR adds fields for cursor width and radius (how rounded the corners of the cursor RRect are) in the TextField class. The default values are configured to match native values.

This PR fixes issue #15571 (https://github.com/flutter/flutter/issues/15571#issuecomment-403182692)

### CODE REVIEW

**Comment by HansMuller** [comment]
Should the cursor be defined by a ShapeDecoration? Assuming that the simple cases are handled without any loss of efficiency, ShapeDecoration (which would have to plumbed all the way through to RenderEditable) would be much more flexible.

File: packages/flutter/lib/src/material/text_field.dart [hunk 1]
```diff
@@ -497,6 +512,9 @@ class _TextFieldState extends State<TextField> with AutomaticKeepAliveClientMixi
         onSelectionChanged: _handleSelectionChanged,
         inputFormatters: formatters,
         rendererIgnoresPointer: true,
+        cursorWidth: widget.cursorWidth,
+        cursorRadius: widget.cursorRadius,
+        cursorColor: widget.cursorColor != null ? widget.cursorColor : Theme.of(context).textSelectionColor,
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/text_field.dart]: cursor.color: widget.cursorColor ?? Theme.of(context).textSelectionColor,

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 2]
```diff
@@ -208,7 +208,7 @@ class EditableText extends StatefulWidget {
     this.onSelectionChanged,
     List<TextInputFormatter> inputFormatters,
     this.rendererIgnoresPointer = false,
-    this.cursorWidth = 1.0,
+    this.cursorWidth = 2.0,
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/editable_text.dart]: This is a backwards incompatible change. Is it really needed?

File: packages/flutter/lib/src/material/text_field.dart [hunk 3]
```diff
@@ -115,6 +115,9 @@ class TextField extends StatefulWidget {
     this.onSubmitted,
     this.inputFormatters,
     this.enabled,
+    this.cursorWidth = 2.0,
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/text_field.dart]: This is a backwards incompatible change. Is it really needed?

File: packages/flutter/test/material/text_field_test.dart [hunk 4]
```diff
@@ -250,6 +250,76 @@ void main() {
...
+  testWidgets('cursor has expected defaults', (WidgetTester tester) async {
+    await tester.pumpWidget(
+        overlay(
+          child: const TextField(
+            cursorRadius: Radius.circular(3.0),
+          ),
+        )
+    );
+
+    final TextField textField =
```
Comment (HansMuller) [inline packages/flutter/test/material/text_field_test.dart]: Just one line for this variable declaration

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 2 above)
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/editable_text.dart]: [per discussion with @sandrasandeep] If the native textfields all use 2.0 wide cursors, and Material doesn't specify a text cursor width, then it makes sense to bite the bullet and make the default 2.0.

So although this is an incompatible change, it's probably best to go though with it now, while Flutter is still in beta.

File: packages/flutter/lib/src/material/text_field.dart (hunk 3 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/text_field.dart]: OK: see https://github.com/flutter/flutter/pull/19317#discussion_r202209668

File: packages/flutter/test/material/text_field_test.dart [hunk 5]
```diff
@@ -250,6 +250,75 @@ void main() {
     await checkCursorToggle();
   });
 
+  testWidgets('cursor has expected defaults', (WidgetTester tester) async {
+    await tester.pumpWidget(
+        overlay(
+          child: const TextField(
+            cursorRadius: Radius.circular(3.0),
```
Comment (HansMuller) [inline packages/flutter/test/material/text_field_test.dart]: It's useful to test that the TextField's cursor has the correct defaults; so don't set the cursorRadius parameter and then verify that it has its default value. The docs say that the default radius is zero. Perhaps it's actually null and that means zero?

File: packages/flutter/test/material/text_field_test.dart [hunk 6]
```diff
@@ -250,6 +250,75 @@ void main() {
...
+          child: new RepaintBoundary(
+            key: const ValueKey<int>(1),
+            child: new TextField(
+              controller: controller,
+              cursorWidth: 15.0,
+            ),
+          ),
+        )
+    );
+    expect(controller.selection.baseOffset, -1);
```
Comment (HansMuller) [inline packages/flutter/test/material/text_field_test.dart]: What do these controller.selection values have to do with checking the cursor's width?

File: packages/flutter/test/material/text_field_test.dart [hunk 7]
```diff
@@ -250,6 +250,75 @@ void main() {
...
+    expect(controller.selection.extentOffset, -1);
+
+    const String testValue = ' ';
+    await tester.enterText(find.byType(TextField), testValue);
+    await skipPastScrollingAnimation(tester);
+
+    await expectLater(
+      find.byKey(const ValueKey<int>(1)),
```
Comment (HansMuller) [inline packages/flutter/test/material/text_field_test.dart]: Wouldn't find.byType(TextField) be adequate?

File: packages/flutter/test/material/text_field_test.dart [hunk 8]
```diff
@@ -250,6 +250,75 @@ void main() {
...
+        )
+    );
+
+    final TextField textField = tester.firstWidget(find.byType(TextField));
+    expect(textField.cursorWidth, 2.0);
+    expect(textField.cursorRadius, const Radius.circular(3.0));
+  });
+
+  testWidgets('cursor layout has correct width', (WidgetTester tester) async {
+    final TextEditingController controller = new TextEditingController();
```
Comment (HansMuller) [inline packages/flutter/test/material/text_field_test.dart]: I don't think an explicit controller is needed for this test.

File: packages/flutter/test/material/text_field_test.dart [hunk 9]
```diff
@@ -250,6 +250,75 @@ void main() {
...
+              controller: controller,
+              cursorWidth: 15.0,
+            ),
+          ),
+        )
+    );
+    expect(controller.selection.baseOffset, -1);
+    expect(controller.selection.extentOffset, -1);
+
+    const String testValue = ' ';
```
Comment (HansMuller) [inline packages/flutter/test/material/text_field_test.dart]: Don't need to define a variable for `testValue`.

File: packages/flutter/test/material/text_field_test.dart [hunk 10]
```diff
@@ -250,6 +250,75 @@ void main() {
...
+    await tester.enterText(find.byType(TextField), testValue);
+    await skipPastScrollingAnimation(tester);
+
+    await expectLater(
+      find.byKey(const ValueKey<int>(1)),
+      matchesGoldenFile('text_field_test.0.0.png'),
+    );
+  }, skip: !Platform.isLinux);
+
+  testWidgets('cursor layout has correct radius', (WidgetTester tester) async {
```
Comment (HansMuller) [inline packages/flutter/test/material/text_field_test.dart]: Same comments as earlier about the test's structure.

File: packages/flutter/test/material/text_field_test.dart [hunk 11]
```diff
@@ -1208,7 +1277,7 @@ void main() {
       editable.getLocalRectForCaret(const TextPosition(offset: 0)).topLeft,
     );
 
-    expect(topLeft.dx, equals(399));
+    expect(topLeft.dx, equals(398.5));
```
Comment (HansMuller) [inline packages/flutter/test/material/text_field_test.dart]: Why did the origin of these tests change? Is the text field's new default cursor width changing TextField's instrinsic width?  I didn't think we expected this PR to introduce any layout changes.

**Review by HansMuller — COMMENTED** [review]
Mostly just some nits. Need to understand what impact on layout this PR will have: will it only the the width of empty textfields or all of them?

File: packages/flutter/test/material/text_field_test.dart (hunk 5 above)
Comment (sandrasandeep) [reply packages/flutter/test/material/text_field_test.dart]: Yes, it turns out it is actually null. I have changed the docs and the tests

File: packages/flutter/test/material/text_field_test.dart (hunk 6 above)
Comment (sandrasandeep) [reply packages/flutter/test/material/text_field_test.dart]: I copied these from another test, but you're right that they don't need to be there. I removed them.

File: packages/flutter/test/material/text_field_test.dart (hunk 7 above)
Comment (sandrasandeep) [reply packages/flutter/test/material/text_field_test.dart]: Yes, changed.

File: packages/flutter/test/material/text_field_test.dart (hunk 11 above)
Comment (sandrasandeep) [reply packages/flutter/test/material/text_field_test.dart]: The new width (1.o -> 2.0) is causing this change.

Comment (HansMuller) [reply packages/flutter/test/material/text_field_test.dart]: Does it only affect empty TextFields?

Comment (sandrasandeep) [reply packages/flutter/test/material/text_field_test.dart]: It looks like it affects textfields with text in them, but the test shows that entering text does not change the layout.

Comment (HansMuller) [reply packages/flutter/test/material/text_field_test.dart]: Sorry I misread this test change: I was assuming that we were checking the TextField's topLeft, not the caret's topLeft. The caret edge moving over a half pixel totally makes sense. Sorry about the confusion.

File: packages/flutter/lib/src/material/text_field.dart [hunk 12]
```diff
@@ -279,6 +282,18 @@ class TextField extends StatefulWidget {
...
+  /// How thick the cursor will be.
+  ///
+  /// Defaults to 2.0.
+  final double cursorWidth;
+
+  /// How rounded the corners of the cursor should be.
+  /// By default, the cursor has a null Radius
+  final Radius cursorRadius;
+
+  /// The color to use when painting the cursor.
```
Comment (xster) [inline packages/flutter/lib/src/material/text_field.dart]: Describe default and null behavior

File: packages/flutter/lib/src/material/theme_data.dart [hunk 13]
```diff
@@ -172,6 +173,7 @@ class ThemeData extends Diagnosticable {
     // Spec doesn't specify a dark theme secondaryHeaderColor, this is a guess.
     secondaryHeaderColor ??= isDark ? Colors.grey[700] : primarySwatch[50];
     textSelectionColor ??= isDark ? accentColor : primarySwatch[200];
+    cursorColor = const Color.fromRGBO(66, 133, 244, 1.0);
```
Comment (xster) [inline packages/flutter/lib/src/material/theme_data.dart]: Isn't this supposed to be derived from the primary swatch? @willlarche

**Review by xster — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/theme_data.dart (hunk 13 above)
Comment (sandrasandeep) [reply packages/flutter/lib/src/material/theme_data.dart]: I will be talking to Will about this tomorrow

**Pull request merged**